wallet/src/de/schildbach/wallet/ui/more/masternode_keys/MasternodeKeyChainFragment.kt

```diff
@@ -64,26 +64,33 @@ class MasternodeKeyChainFragment : Fragment(R.layout.fragment_masternode_key_cha
                 MasternodeKeyType.OPERATOR -> R.string.masternode_key_type_operator
                 MasternodeKeyType.PLATFORM -> R.string.masternode_key_type_platform
                 else -> throw IllegalArgumentException("invalid masternode key type")
-            },
+            }
         )
         masternodeKeyChainAdapter = MasternodeKeyChainAdapter(
-            viewModel.getKeyChainInfo(masternodeKeyType),
+            viewModel.getKeyChainInfo(masternodeKeyType, false),
             viewModel.getKeyUsage(),
             { handleCopyAddress(it) },
-            { key, position -> handleDecryptKey(key, position) },
+            { key, position -> handleDecryptKey(key, position) }
         )
         binding.keyList.adapter = masternodeKeyChainAdapter
         binding.keyList.layoutManager = LinearLayoutManager(activity)
         binding.keyList.setHasFixedSize(true)
         masternodeKeyChainAdapter.notifyDataSetChanged()
 
         binding.addMasternodeKey.setOnClickListener {
-            val position = viewModel.addKey(masternodeKeyType)
-            masternodeKeyChainAdapter.addKey(position)
-            (binding.keyList.layoutManager as LinearLayoutManager).scrollToPositionWithOffset(
-                position,
-                0,
-            )
+            lifecycleScope.launch {
+                val position = viewModel.addKey(masternodeKeyType)
+                masternodeKeyChainAdapter.addKey(position)
+                (binding.keyList.layoutManager as LinearLayoutManager).scrollToPositionWithOffset(
+                    position,
+                    0
+                )
+            }
+        }
+        viewModel.newKeysFound.observe(viewLifecycleOwner) { isAdded ->
+            if (isAdded == true) {
+                masternodeKeyChainAdapter.notifyDataSetChanged()
+            }
```

Comment:
Here the we can react to the event and update the UI.

Reply:
The usage of LiveData is fine here since the fragment observes an asynchronous event. However, an improvement can be made to eliminate the `notifyDataSetChanged` calls, which are not very performant on large lists since it forces the `RecycleView` to redraw everything.

Instead of observing "listWasChanged" event which is represented by `newKeysFound` livedata here and then notifying the adapter, the fragment can observe the list itself and then provide it to the adapter. 
The viewModel can hold a `LiveData<List<MasternodeKeyInfo>>` or another `data class` that holds all the necessary info for each row. This list is updated in the viewModel when necessary, and the fragment submits it to the adapter when the list is changed, for example in the `TransactionGroup` classes:

```
// In the viewModel
private val _transactions = MutableLiveData<List<TransactionRowView>>()
val transactions: LiveData<List<TransactionRowView>>
    get() = _transactions

// In the fragment
viewModel.transactions.observe(viewLifecycleOwner) { transactions ->
    adapter.submitList(transactions)
}
```

Then the adapter can derive from `ListAdapter` to have a `DiffCallback`, like in the `TransactionAdapter`.
The `DiffCallback` will have `areItemsTheSame` method which compares unique item IDs (for example, the `masternodeKey` field)  and `areContentsTheSame` method which will compare all of the contents (with `data class` it's easy - just check if they are equal). 

This will ensure that the `RecycleView` only redraws the changed rows and the changed content in those rows. If it sees that a new unique ID is appeared in the list (or removed), it will add/remove a new viewHolder to the list. If it sees that no items are added/removed, but some item have their contents changed (because the `data class` have changed fields), it will update the content in that row specifically.

Another advantage of this approach is that the adapter doesn't need to know much about the specifics of our MasternodeKeys feature, and only interacts with an abstract representation of the row view itself.